CodeGeneration/Sources/SyntaxSupport/Node.swift

```diff
@@ -55,6 +55,8 @@ public class Node: NodeChoiceConvertible {
   /// function that should be invoked to create this node.
   public let parserFunction: TokenSyntax?
 
+  public let disableSameTypeForUniqueChoice: Bool
```

Comment:
Please correct me if I’m wrong but I think we only support this for syntax collections, right? If so, I think it would be a cleaner design to modify [`Node.Data.collection`](https://github.com/swiftlang/swift-syntax/blob/d145cb2a6efb5c8d45cd5df5f494a89c93c90758/CodeGeneration/Sources/SyntaxSupport/Node.swift#L27) to

```swift
fileprivate enum SyntaxCollectionElement {
  /// The syntax collection can only have a single kind of syntax node as its element.
  case singleOption(SyntaxNodeKind)
  /// Create a `SyntaxChildChoices` enum for the collection’s elements.
  case choices([SyntaxNodeKind])
}

case collection(SyntaxCollectionElement)
```

And with that `CollectionNode` can gain

```swift
var generateSyntaxChoicesEnum: Bool {
    switch node.data {
    case .layout:
      preconditionFailure("NodeLayoutView must wrap a Node with data `.collection`")
    case .collection(.singleOption(_)): 
      return false
    case .collection(.choices(_)): 
      return true
    }
} 
```

That way we don’t need to repeat the `> 1 || disableSameTypeForUniqueChoice` check everywhere.